tests/e2e/tidbcluster/tls.go

```diff
@@ -0,0 +1,271 @@
+	"github.com/pingcap/tidb-operator/pkg/util"
+	"github.com/pingcap/tidb-operator/tests/e2e/util/portforward"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+)
+
+var tidbIssuerTmpl = `
+apiVersion: cert-manager.io/v1alpha2
+kind: Issuer
+metadata:
+  name: {{ .ClusterName }}-selfsigned-ca-issuer
+  namespace: {{ .Namespace }}
+spec:
+  selfSigned: {}
+---
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: {{ .ClusterName }}-ca
+  namespace: {{ .Namespace }}
+spec:
+  secretName: {{ .ClusterName }}-ca-secret
+  commonName: "TiDB CA"
+  isCA: true
+  issuerRef:
+    name: {{ .ClusterName }}-selfsigned-ca-issuer
+    kind: Issuer
+---
+apiVersion: cert-manager.io/v1alpha2
+kind: Issuer
+metadata:
+  name: {{ .ClusterName }}-tidb-issuer
+  namespace: {{ .Namespace }}
+spec:
+  ca:
+    secretName: {{ .ClusterName }}-ca-secret
+`
+
+var tidbCertificatesTmpl = `
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: {{ .ClusterName }}-tidb-server-secret
+  namespace: {{ .Namespace }}
+spec:
+  secretName: {{ .ClusterName }}-tidb-server-secret
+  duration: 8760h # 365d
+  renewBefore: 360h # 15d
+  organization:
+    - PingCAP
+  commonName: "TiDB Server"
+  usages:
+    - server auth
+  dnsNames:
+    - "{{ .ClusterName }}-tidb"
+    - "{{ .ClusterName }}-tidb.{{ .Namespace }}"
+    - "*.{{ .ClusterName }}-tidb"
+    - "{{ .ClusterName }}-tidb.{{ .Namespace }}.svc"
+    - "*.{{ .ClusterName }}-tidb.{{ .Namespace }}"
+    - "*.{{ .ClusterName }}-tidb.{{ .Namespace }}.svc"
+  ipAddresses:
+    - 127.0.0.1
+    - ::1
+  issuerRef:
+    name: {{ .ClusterName }}-tidb-issuer
+    kind: Issuer
+    group: cert-manager.io
+---
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: {{ .ClusterName }}-tidb-client-secret
+  namespace: {{ .Namespace }}
+spec:
+  secretName: {{ .ClusterName }}-tidb-client-secret
+  duration: 8760h # 365d
+  renewBefore: 360h # 15d
+  organization:
+    - PingCAP
+  commonName: "TiDB Client"
+  usages:
+    - client auth
+  issuerRef:
+    name: {{ .ClusterName }}-tidb-issuer
+    kind: Issuer
+    group: cert-manager.io
+`
+
+type tidbClusterTmplMeta struct {
+	Namespace   string
+	ClusterName string
+}
+
+func installCertManager(cli clientset.Interface) error {
+	cmd := "kubectl apply -f /cert-manager.yaml --validate=false"
+	if data, err := exec.Command("sh", "-c", cmd).CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to install cert-manager %s %v", string(data), err)
+	}
+
+	err := e2epod.WaitForPodsRunningReady(cli, "cert-manager", 3, 0, 10*time.Minute, nil)
+	if err != nil {
+		return err
+	}
+
+	// It may take a minute or so for the TLS assets required for the webhook to function to be provisioned.
+	time.Sleep(time.Minute)
```

Comment:
can we wait for the webhook is ready?

Reply:
agreed

Reply:
OK, I will do it in the next pr.

Reply:
open an issue for this?

actually, when the webhook pod is ready, the webhook TLS secret is created. if the API request may still fail, we can wait for the mutating/validating webhook configurations CA bundles to be injected

Reply:
Opened an issue: https://github.com/pingcap/tidb-operator/issues/2247